pupil_src/shared_modules/background_task/task.py

```diff
@@ -0,0 +1,175 @@
+            if not should_continue:
+                return
+
+    def _handle_signal_normally(self, signal):
+        if isinstance(signal, _TaskCompletedSignal):
+            self.on_completed(signal.return_value)
+            return False
+        elif isinstance(signal, _TaskExceptionSignal):
+            # Unfortunately, background exceptions raised in the foreground don't
+            # have a proper traceback.
+            # If you are debugging an exception, you can print datum.traceback to
+            # get the traceback in the other process. Just uncomment:
+            # print(signal.traceback)
+            # If this happens often, we can consider using tblib to send tracebacks
+            # to the foreground (see https://stackoverflow.com/a/26096355)
+            self.on_exception(signal.exception)
+            return False
+        elif isinstance(signal, _TaskYieldSignal):
+            self.on_yield(signal.datum)
+            return True
+        else:
+            raise ValueError(
+                "Received unknown signal {} from background " "process".format(signal)
+            )
+
+    def _handle_signal_if_canceled(self, signal):
+        if (
+            isinstance(signal, _TaskCanceledSignal)
+            or isinstance(signal, _TaskCompletedSignal)
+            or isinstance(signal, _TaskExceptionSignal)
+        ):
+            self.on_canceled_or_killed()
+            return False
+        elif isinstance(signal, _TaskYieldSignal):
+            return True
+        else:
+            raise ValueError(
+                "Received unknown signal {} from background " "process".format(signal)
+            )
+
+
+class BackgroundGeneratorFunction(BackgroundTaskBase):
+    def get_process(self, name, generator_function, args, kwargs, pipe_send, patches):
+        wrapper_kwargs = {
+            "pipe_send": pipe_send,
+            "should_terminate_flag": self._should_terminate_flag,
+            "generator_function": generator_function,
+            "args": args,
+            "kwargs": kwargs,
+            "patches": patches,
+        }
+
+        return mp.Process(target=_generator_wrapper, name=name, kwargs=wrapper_kwargs)
+
+
+def _generator_wrapper(
+    pipe_send, should_terminate_flag, generator_function, args, kwargs, patches
+):
+    """Executed in background, pipes results to foreground"""
+    try:
+        for patch in patches:
+            patch.apply()
+        for datum in generator_function(*args, **kwargs):
+            if should_terminate_flag.value:
+                pipe_send.send(_TaskCanceledSignal())
+                return  # will also trigger "finally"
+            pipe_send.send(_TaskYieldSignal(datum))
+    except Exception as e:
+        import traceback
+
+        pipe_send.send(_TaskExceptionSignal(e, traceback.format_exc()))
+    else:
+        pipe_send.send(_TaskCompletedSignal(return_value=None))
+    finally:
+        pipe_send.close()
+
+
+class BackgroundRoutine(BackgroundTaskBase):
```

Comment:
Could it make sense to overwrite extend `cancel_gracefully` and throw an error since a routine can not be cancelled gracefully, respectively it would not make sense to call this?

Reply:
Well, it makes sense to cancel_gracefully() routines. Even if they continue normally, no on_completed will be called if it was canceled, so you can safely just start the task again with different parameters.

Additionally, now you can cancel routines by passing shared memory and looking into the `should_terminate_flag` (see latest commit).

Reply:
Makes sense!